### PR TITLE
Feat/sm120 backend

### PR DESCRIPTION
**Pull request by webcodes-cz**
## 📌 Description

This PR adds first-class **SM120 (Blackwell / RTX 5090)** backend support across FlashInfer’s FP4 quantization and fused-MoE paths, plus a set of robustness fixes discovered while bringing up GPT-OSS-20B on RTX 5090.

### What’s included

- **Auto SM backend selection** with an env override
  - New `_resolve_backend("auto")` chooses **"120"**, **"100"**, or **"90"** from CUDA compute capability.
  - `FLASHINFER_FORCE_SM` allows forcing **120/120a/100/90/90a**.
- **SM120 JIT support**
  - Adds `sm120_nvcc_flags` & `sm120a_nvcc_flags` in JIT.
  - New builders:
    - `gen_fp4_quantization_sm120_module()`
    - `gen_cutlass_fused_moe_sm120_module()`
  - `flashinfer/jit/v1.2.7.post1/__init__.py` stub to avoid hard-coded SM100 imports and let JIT pick the right arch.
- **MXFP4 hardening**
  - **K%4 alignment**: scale-factor shuffling now pads K to a multiple of 4; downstream **reshape** uses padded shapes (fixes invalid reshape when K=90→92).
  - **Safer A-shuffle**: `shuffle_matrix_a` now asserts M%128 (callers must pad) instead of silently padding, removing a risky implicit behavior.
- **MoE path**
  - Fused-MoE core picks SM120 backend when available; accepts `FLASHINFER_SM120_VARIANT=120|120a` (default **120a**) for toolchains that differ on the “a” suffix support.
- **Version metadata**
  - `__version_extra__`: `VERSION_SUFFIX = "+sm120-local"` for easier local artifact identification during bring-up.

---

## 🔍 Related Issues

- Blackwell/5090 bring-up (internal)
- K%4 scale-shuffle & reshape mismatch on FP4 (internal)

---

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` and the hooks.
- [ ] I have run `pre-commit run --all-files` and fixed issues.

### 🧪 Tests

- [ ] Unit tests for `_resolve_backend()` (env override + auto by capability).
- [ ] FP4 quantization smoke test on SM120 (verifies JIT selects 120/120a and builds).
- [ ] MXFP4 scale-shuffle round-trip with **K not divisible by 4** (no reshape error).
- [ ] Fused-MoE SM120 smoke test (builds once, loads cached op, returns correct shapes).
- [ ] Backward compatibility on SM100/SM90 (no behavior change).

---

## 🧰 Implementation Notes

- **Files (high level)**
  - `flashinfer/jit/core.py`: add `sm120_nvcc_flags`, `sm120a_nvcc_flags`.
  - `flashinfer/jit/__init__.py`: export new flags.
  - `flashinfer/jit/v1.2.7.post1/__init__.py`: empty stub to prevent SM100-only imports.
  - `flashinfer/fp4_quantization.py`:
    - add `gen_fp4_quantization_sm120_module()`.
    - introduce `_resolve_backend()` and default `get_fp4_quantization_module(backend="auto")`.
    - K%4 padding + padded-shape reshape fix.
    - safer `.to(torch.uint8)` conversions.
  - `flashinfer/fused_moe/core.py`:
    - add `gen_cutlass_fused_moe_sm120_module()`.
    - `_resolve_backend()` + default `backend="auto"`.
  - `flashinfer/utils.py`:
    - comment on K%4 assert clarifies caller-padding contract.
  - `flashinfer/__version_extra__.py`: `+sm120-local`.

- **Environment variables**
  - `FLASHINFER_FORCE_SM=120|120a|100|90|90a`
    - Forces backend selection regardless of device capability.
  - `FLASHINFER_SM120_VARIANT=120a|120`
    - Selects **sm_120a** (default) vs **sm_120** for toolchain portability.

---

## 📝 Backward Compatibility

- Default behavior remains unchanged on SM100/SM90: auto detection picks the existing paths; env override is opt-in.
- No public API changes.

---

## 📊 Validation (manual smoke)

```bash
# Clear JIT caches to observe correct target
rm -rf ~/.cache/flashinfer ~/.cache/torch_extensions

# Run a minimal FP4 quantization op on RTX 5090
export FLASHINFER_FORCE_SM=120   # optional; otherwise auto-detects 120
python - <<'PY'
import torch
from flashinfer.fp4_quantization import get_fp4_quantization_module
m = get_fp4_quantization_module(backend="auto")
print("Loaded:", m)
PY
# Expect: build once under ~/.cache/flashinfer/120/... then reuse.

### CODE REVIEW

**Comment by yongwww** [comment]
@webcodes-cz  thanks for the effort! Some changes might have been handled in https://github.com/flashinfer-ai/flashinfer/pull/1608

File: flashinfer/fp4_quantization.py [hunk 1]
```diff
@@ -71,6 +73,17 @@ def gen_fp4_quantization_sm90_module() -> JitSpec:
     return gen_fp4_quantization_module(sm90a_nvcc_flags, "90")
 
 
+def gen_fp4_quantization_sm120_module() -> JitSpec:
```
Comment (yongwww) [inline flashinfer/fp4_quantization.py:76]: seems the fp4 quant was handled in https://github.com/flashinfer-ai/flashinfer/pull/1609.

Comment (webcodes-cz) [reply flashinfer/fp4_quantization.py:76]: great, you did it. I was just tired of waiting.. Closing pull req.

Comment (yongwww) [reply flashinfer/fp4_quantization.py:76]: Thanks, @webcodes-cz, for the great effort! Please feel free to submit any follow-up PRs or fixes.
